tests/integration/devfile/cmd_dev_test.go

```diff
@@ -1393,6 +1393,91 @@ var _ = Describe("odo dev command tests", func() {
 		})
 	})
 
+	When("running an alternative non-default run command", func() {
+		BeforeEach(func() {
+			helper.CopyExampleDevFile(
+				filepath.Join("source", "devfiles", "nodejs", "devfile-with-two-run-commands.yaml"),
+				filepath.Join(commonVar.Context, "devfile.yaml"))
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), commonVar.Context)
+		})
+
+		It("should error out if called with an invalid command", func() {
+			output := helper.Cmd("odo", "dev", "--random-ports", "--run-command", "run-command-does-not-exist").ShouldFail().Err()
+			Expect(output).To(ContainSubstring("no run command with name \"run-command-does-not-exist\" found in Devfile"))
+		})
+
+		It("should error out if called with a command of another kind", func() {
+			//devbuild is a valid build command, not a run command
+			output := helper.Cmd("odo", "dev", "--random-ports", "--run-command", "devbuild").ShouldFail().Err()
+			Expect(output).To(ContainSubstring("no run command with name \"devbuild\" found in Devfile"))
+		})
+
+		testForCmd := func(devfileCmd string, checkFunc func(stdout, stderr string)) {
```

Comment:
> We are checking only for the messages being printed by odo on stdout. Not so long back, we discussed that we shouldn't be checking too much of such messages in the output. Instead, if we could test that the functionality is working as expected, that would be better.
> 
> For example, in tests checking for success, we could check for the state of things on the cluster instead of comparing with stdout. In tests checking for failure, it makes sense to be validating against the messages printed on stdout/stderr.
> 
> If you look at the previous tests for successful run of `odo dev`, you would see that are checking if expected output shows up when you `curl` a port that's forwarded to k8s cluster instead of chcking if odo prints specific thing on stdout.
> 
> Finally, I'm not against checking stdout, but IMO there might be better avenues to test successful runs than just checking stdout.

Thanks for the detailed explanations.. This is exactly what I've tried to do in these tests. If you scroll a few lines below, you will notice that besides checking the output, it also performs additional checks, like checking the port by issuing an HTTP request (actually, I took inspiration from the existing tests.). 

https://github.com/rm3l/odo/blob/5775-odo-dev-add-flag-to-run-non-default-run-command/tests/integration/devfile/cmd_dev_test.go#L1420-L1439

The `checkFunc` function parameter allows each individual test case to perform additional checks. For example, it allowed to perform negative assertions on the output (to make it did not execute any unwanted run command), but also to check for example that the custom run command did create the expected folder in the remote container:
https://github.com/rm3l/odo/blob/5775-odo-dev-add-flag-to-run-non-default-run-command/tests/integration/devfile/cmd_dev_test.go#L1447-L1465

Reply:
The checks for string outputs is happening in `By` which I stepped through after reading your response. Those calls are harmless (in terms of time they take vs. an `It` block). So, I guess, we are good. :)